spline-shadow/persistence/src/main/scala/za/co/absa/spline/persistence/Persister.scala

```diff
@@ -17,235 +17,124 @@
-        case _ => Transformation(name, expression, outputSchema, _key)
-      }})
-
-  private def createEncodedOperations(dataLineage: DataLineage) = {
-    val encoderWrite: ObjectEncoder[Write] = deriveEncoder[Write]
-    val encoderRead: ObjectEncoder[Read] = deriveEncoder[Read]
-    val encoderTransformation: ObjectEncoder[Transformation] = deriveEncoder[Transformation]
-    createOperations(dataLineage)
-      .map {
-        case read: Read => encoderRead(read)
-        case transformation: Transformation => encoderTransformation(transformation)
-        case write: Write => encoderWrite(write)
-      }
-      .map(_.noSpaces)
   }
 
-  private def createFollows(dataLineage: DataLineage) = {
-    // Operation inputs and outputs ids may be shared across already linked lineages. To avoid saving linked lineages or
-    // duplicate indexes we need to not use these.
-    val outputToOperationId = dataLineage
-      .operations
-      .map(o => (o.mainProps.output, o.mainProps.id))
+  private def queryExistingToKey(uris: Seq[String]): Map[String, String] = {
+    val urisList = uris
+      .map(uri => "\"" + uri + "\"")
+      .mkString(", ")
+    val query = s"for ds in dataSource filter ds.uri in [$urisList] return ds"
+    val result = db.query(query, classOf[String])
+    result
+      .asInstanceOf[java.util.Iterator[String]]
+      .asScala
+      .map(s => {
+        val json = parse(s).right.get
+        deriveDecoder[DataSource]
+          .decodeJson(json)
+          .right.get
+      })
+      .map(ds => ds.uri -> ds._key.get)
       .toMap
-    dataLineage.operations.iterator.toIterable
-      .flatMap(op =>
-        op.mainProps.inputs
-          .flatMap(outputToOperationId.get)
-          .map(opId => Follows(
-            "operation/" + op.mainProps.id.toString,
-            "operation/" + opId.toString,
-            Some(op.mainProps.id.toString + '-' + opId.toString)))
-        )
-      .map(deriveEncoder[Follows].apply)
-      .map(_.noSpaces)
   }
 
-  private def findOutputSchema(dataLineage: DataLineage, operation: splinemodel.op.Operation): Schema = {
-    val metaDataset: MetaDataset = dataLineage.datasets.find((dts: MetaDataset) => dts.id == operation.mainProps.output)
-      .getOrElse(throw new IllegalArgumentException(
-        s"Operation output id ${operation.mainProps.output} not found in datasets of dataLineage ${dataLineage.id}"))
-    val attributes = metaDataset.schema.attrs.map(attrId => {
-      val attribute = dataLineage.attributes.find(_.id == attrId)
-        .getOrElse(throw new IllegalArgumentException(
-          s"MetaDataset ${metaDataset.id} contains Attribute $attrId " +
-            s"that is not available in Datalineage#attributes of ${dataLineage.id}."))
-      Attribute(attribute.name, attribute.dataTypeId.toString)
-    })
-    Schema(attributes)
+  private def filterUriToNewKey(uris: Seq[String], uriToExistingKey: Map[String, String]) = {
+    uris
```

Comment:
can be simplified. E.g. `(uris.toSet -- uriToExistingKey.keys).map(_ -> randomUUID.toString)`